### PR TITLE
refactor:  refactor output format with FieldEncoders

### PR DESCRIPTION
**Pull request by youngsofun**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

change 

```
trait TypeSerializer {
   fn write_tsv();
   fn write_csv();
   fn write_json();
  ...
}
```
to 

```
trait FieldEncoder {
   fn write_string(s: &StringSerializer );
   fn write_int(s: IntSerializer);
  ...
}

struct TSVFieldEncoder{}

impl FieldEncoder for TSVFieldEncoder {
}
```

1. more readable to understand a format.
3. more clear nested format,  e.g. CSV/NDJson Encoder `has a`  Values Encoder to encode nested types. 
4. much clear call to self, especially when mixed with Inherited functions.
5. correct deps:  crate `format` now depends on `datavalues`, Instead of the other way around.
6. XXXFieldEncoder contains settings need by  format XXX, no global shared `FormatSettings` 

todo next(not this pr):

-  funcs like `write_tsv` are still used in a few other places, which will be resolved in next pr.
-  we can get rid of FormatSetting after a likely refactor for input formats.

### CODE REVIEW

**Comment by youngsofun** [comment]
@BohuTANG why only checks runned?

**Comment by BohuTANG** [comment]
@mergify update

**Comment by BohuTANG** [comment]
@Xuanwo mergify looks not working

**Review by b41sh — APPROVED** [review]

File: src/query/datavalues/src/types/serializations/array.rs [hunk 1]
```diff
@@ -22,8 +22,8 @@ use crate::serializations::write_json_string;
 
 #[derive(Clone)]
 pub struct ArraySerializer<'a> {
-    offsets: &'a [i64],
-    inner: Box<TypeSerializerImpl<'a>>,
+    pub offsets: &'a [i64],
```
[the remark this answers is not in the document]
Comment (youngsofun) [reply src/query/datavalues/src/types/serializations/array.rs:25]: I am not very sure about it in  this case, 
but XXXSerializer will be migrated to  `new expression`,  what about applying the style then?

[the remark this answers is not in the document]
Comment (Xuanwo) [reply src/query/datavalues/src/types/serializations/array.rs:25]: LGTM

**Review by Xuanwo — APPROVED** [review]

**Review by sundy-li — APPROVED** [review]

**Review by sundy-li — APPROVED** [review]

**Pull request merged**